### PR TITLE
fix(dijkstra): Added negative fixtures for duplicate witness and redeemer encodings

### LINKED ISSUE #1799: dijkstra: add negative fixtures for duplicate witness and redeemer encodings
## Summary

Dijkstra duplicate witness/redeemer hardening needs explicit negative-fixture
coverage once upstream fixtures are available.

Current code state:
- Shared CBOR decode mode uses `DupMapKeyEnforcedAPF`, so duplicate map keys are rejected globally.
- `DijkstraRedeemers.UnmarshalCBOR()` requires map encoding and decodes through `cbor.DecodeGeneric`, so duplicate redeemer keys should fail at decode time.
- Dijkstra witness-set fields such as vkeys, native scripts, Plutus scripts, datums, and sub-transactions use `cbor.SetType[...]`, which currently decodes set contents as arrays and preserves members.
- Current Dijkstra tests cover positive witness/redeemer behavior and extraneous redeemers, but not upstream negative CBOR fixtures for duplicate set members or duplicate redeemer map keys.

## What needs to happen

1. Add upstream negative fixtures in `ouroboros-mock` when available, or add synthetic local tests until fixtures exist.
2. Verify duplicate Dijkstra redeemer map keys fail during decode, before rule evaluation.
3. Verify duplicate witness-set members and duplicate sub-transaction set members follow the ledger/spec expectation for Dijkstra sets.
4. If duplicates should be rejected but `cbor.SetType` permits them, add a Dijkstra decode or validation guard without breaking older era encodings.

## Verify

```bash
go test ./ledger/dijkstra/... -run 'Duplicate|Redeemer|Witness' -v
go test ./internal/test/conformance/... -v
```

## Notes

This issue is intentionally scoped to Dijkstra hardening and fixture coverage.
It should not claim the shared duplicate-map-key decoder is missing; that guard
is already present in `cbor/decode.go`.

### PR DESCRIPTION
**Pull request by arepala-uml**
1. Added `CheckForDuplicates()` error to `cbor.SetType[T]` — encodes each item to CBOR bytes and checks for byte-level duplicates; no-ops on untagged arrays so pre-Dijkstra encodings are unaffected
2. Extended `DijkstraTransactionWitnessSet.UnmarshalCBOR `to call the check for duplicates on all eight witness-set fields (`vkeys, native scripts, bootstrap witnesses, Plutus V1–V4 scripts, datums`)
3. Extended `DijkstraTransactionBody.UnmarshalCBOR` to call the check for duplicates on the three SetType body fields (`TxCollateral, TxReferenceInputs, TxSubTransactions`)
4. Added four negative tests covering duplicate redeemer map key (existing `DupMapKeyEnforcedAPF`), duplicate tagged vkey witness, duplicate tagged sub-transaction, and duplicate untagged vkey witness.
5. Wired duplicate-member validation into Conway transaction body decoding for tagged `TxInputs`, since Dijkstra reuses `ConwayTransactionInputSet` for its input set decode path.

Closes #1799 

---
## Summary by cubic
Reject duplicate members in tag-258 CBOR sets across Dijkstra and Conway witness sets, transaction bodies, guards, and nested sub-transactions. Untagged arrays remain unchanged. Fulfills #1799.

- **Bug Fixes**
  - Added `CheckForDuplicates()` to `cbor.SetType[T]` and `ConwayTransactionInputSet`; tagged sets only; returns "duplicate member in set".
  - Enforced duplicate rejection on decode in Dijkstra and Conway: `DijkstraTransactionWitnessSet.UnmarshalCBOR`, `DijkstraTransactionBody.UnmarshalCBOR`, `DijkstraSubTransactionBody.UnmarshalCBOR`, `DijkstraGuards.UnmarshalCBOR`, plus `ConwayTransactionWitnessSet.UnmarshalCBOR` and `ConwayTransactionBody.UnmarshalCBOR` for witness fields and body sets (`TxInputs`, `TxCollateral`, `TxRequiredSigners`, `TxReferenceInputs`, and `TxSubTransactions` where applicable).
  - Expanded tests to cover duplicate tagged vkeys, inputs, body sets, sub-transactions and nested sub-tx inputs/reference inputs, guards, and redeemer map keys; confirmed untagged arrays remain allowed.

<sup>Written for commit ddc411766a475a801a44b7e5f67edef2dbbbce5b. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/blinklabs-io/gouroboros/pull/1848?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved validation when decoding certain transaction and witness data, now rejecting duplicate members in tagged set fields.
  * Duplicate map keys in redeemer data are now detected and rejected.
  * Untagged list-based encodings continue to be accepted for compatibility.

* **Tests**
  * Added coverage for duplicate-member rejection and compatibility with untagged encodings.

### CODE REVIEW

**Comment by arepala-uml** [comment]
@cubic-dev-ai

File: ledger/dijkstra/dijkstra.go [hunk 1]
```diff
@@ -907,6 +907,19 @@ func (b *DijkstraTransactionBody) UnmarshalCBOR(cborData []byte) error {
 	if _, err := cbor.Decode(cborData, &tmp); err != nil {
 		return err
 	}
+	// Reject duplicate members in any tag-258 set field on the transaction body.
+	type duplicateChecker interface {
+		CheckForDuplicates() error
+	}
+	for _, c := range []duplicateChecker{
```
Comment (wolf31o2) [inline ledger/dijkstra/dijkstra.go:920]: This list still misses the guards path. `TxGuards` decodes through `DijkstraGuards.UnmarshalCBOR` as a `cbor.SetType`, but that code returns without calling `CheckForDuplicates()`. I was able to decode a body with duplicate tagged guard members without an error. Please apply the duplicate check there too, for both credential and key-hash guard forms.

File: ledger/dijkstra/dijkstra.go [hunk 2]
```diff
@@ -907,6 +907,19 @@ func (b *DijkstraTransactionBody) UnmarshalCBOR(cborData []byte) error {
...
+		&tmp.TxCollateral,
+		&tmp.TxReferenceInputs,
+		&tmp.TxSubTransactions,
```
Comment (wolf31o2) [inline ledger/dijkstra/dijkstra.go:924]: Checking duplicates among `TxSubTransactions` is useful, but it does not catch duplicate sets inside each sub-transaction body. `DijkstraSubTransactionBody.TxReferenceInputs` is also a `cbor.SetType`, and its `UnmarshalCBOR` currently does no duplicate check, so duplicate tagged reference inputs inside a sub-transaction still pass decode.

File: ledger/dijkstra/dijkstra_test.go [hunk 3]
```diff
@@ -331,6 +331,73 @@ func TestDijkstraBlockRoundTripWithBodyHash(t *testing.T) {
...
+		0x82,                         // array(2)
+		0x82, 0x41, 0x01, 0x41, 0x02, // VkeyWitness{[0x01], [0x02]}
+		0x82, 0x41, 0x01, 0x41, 0x02, // duplicate
+	}
+	var ws DijkstraTransactionWitnessSet
+	err := ws.UnmarshalCBOR(dupCbor)
+	require.ErrorContains(t, err, "duplicate member in set")
+}
+
+func TestDijkstraTransactionBodyRejectsDuplicateSubTransaction(t *testing.T) {
```
Comment (wolf31o2) [inline ledger/dijkstra/dijkstra_test.go:370]: Can we add negative tests for duplicate tagged guards and duplicate tagged sub-transaction reference inputs as well? Those two cases currently pass decode, so the new tests make the coverage look broader than the implementation actually is.

**Review by wolf31o2 — CHANGES_REQUESTED** [review]
Thanks for adding the generic set duplicate check. I do not think this is ready to merge yet: Dijkstra still has tag-258 set decode paths that do not call the new guard, so malformed CBOR with duplicate members can still get through. I left the specific cases inline. Local `make test` and `make lint` both pass, but these missing decode paths need to be closed and covered before merge.

File: ledger/dijkstra/dijkstra.go (hunk 1 above)
Comment (arepala-uml) [reply ledger/dijkstra/dijkstra.go:920]: I have fixed it where `DijkstraGuards.UnmarshalCBOR` now calls CheckForDuplicates() for both decode branches such as credential guards and key-hash guards. Also, added negative tests for both forms through `DijkstraTransactionBody.UnmarshalCBOR`, so duplicate tagged guard members now fail decode with duplicate member in set.

File: ledger/dijkstra/dijkstra.go (hunk 2 above)
Comment (arepala-uml) [reply ledger/dijkstra/dijkstra.go:924]: I have fixed it where `DijkstraSubTransactionBody.UnmarshalCBOR` checks duplicate members in its nested tag-258 set fields, including TxReferenceInputs. Also, added a regression test that decodes a top-level transaction body with a unique TxSubTransactions set, where the contained sub-transaction has duplicate tagged reference inputs.

File: ledger/dijkstra/dijkstra_test.go (hunk 3 above)
Comment (arepala-uml) [reply ledger/dijkstra/dijkstra_test.go:370]: As above mentioned, I have added negative decode tests for both missed cases such as duplicate tagged guards (credential guard form, key-hash guard form) and duplicate tagged sub-transaction reference inputs.

So, these three tests go through `DijkstraTransactionBody.UnmarshalCBOR` and it covers the real transaction body decode path where duplicate members now fail with duplicate member in set.

**Comment by arepala-uml** [comment]
> Thanks for adding the generic set duplicate check. I do not think this is ready to merge yet: Dijkstra still has tag-258 set decode paths that do not call the new guard, so malformed CBOR with duplicate members can still get through. I left the specific cases inline. Local `make test` and `make lint` both pass, but these missing decode paths need to be closed and covered before merge.

Understood, I have added the missing duplicate checks for guards, top-level inputs, and nested sub-transaction inputs/reference inputs. Also, I have added Conway TxInputs into the same check because it uses ConwayTransactionInputSet and once that type got CheckForDuplicates(), Conway’s own body decode needed to call it too. That is the reason, I have modified it in conway and please review it.

**Comment by arepala-uml** [comment]
@cubic-dev-ai

**Comment by arepala-uml** [comment]
I have added a table below where we handled duplicate-member rejection for all tag-258 CBOR sets across Dijkstra and conway.

Type | Field | Added in
-- | -- | --
DijkstraTransactionBody | TxInputs, TxCollateral, TxReferenceInputs, TxSubTransactions | ledger/dijkstra/dijkstra.go
DijkstraSubTransactionBody | TxInputs, TxReferenceInputs | ledger/dijkstra/dijkstra.go
DijkstraTransactionWitnessSet | all 8 witness-set fields | ledger/dijkstra/dijkstra.go
DijkstraGuards | credentials, keyHashes | ledger/dijkstra/dijkstra.go
ConwayTransactionBody | TxInputs, TxCollateral, TxRequiredSigners, TxReferenceInputs | ledger/conway/conway.go
ConwayTransactionWitnessSet | all 7 witness-set fields | ledger/conway/conway.go

**Review by wolf31o2 — APPROVED** [review]
